### PR TITLE
Add httpapi password change driver to connect to generic HTTP/HTTPS APIs

### PR DESCRIPTION
**Pull request by davidc**

### CODE REVIEW

File: plugins/password/drivers/httpapi.php [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    /**
+     * This method is called from roundcube to change the password
+     *
+     * roundcube already validated the old password so we just need to change it at this point
+     *
+     * @author David Croft
+     * @param string $curpass Current password
+     * @param string $newpass New password
+     * @param string $username Login username
+     * @returns int PASSWORD_SUCCESS|PASSWORD_ERROR|PASSWORD_CONNECT_ERROR
```
Comment (alecpl) [inline plugins/password/drivers/httpapi.php:47]: It should be `@return`.

File: plugins/password/drivers/httpapi.php [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	$var_user = $rcmail->config->get('password_httpapi_var_user');
+	$var_curpass = $rcmail->config->get('password_httpapi_var_curpass');
+	$var_newpass = $rcmail->config->get('password_httpapi_var_newpass');
+	$expect = $rcmail->config->get('password_httpapi_expect');
+
+	// Initialise curl options
+
+        $curl = curl_init();
+
+        curl_setopt($curl, CURLOPT_CONNECTTIMEOUT , 5);
```
Comment (alecpl) [inline plugins/password/drivers/httpapi.php:67]: Spacing incorrect.

File: plugins/password/drivers/httpapi.php [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+	if ($ssl_verify) {
+	    curl_setopt($curl, CURLOPT_SSL_VERIFYHOST , 2);
+	    curl_setopt($curl, CURLOPT_SSL_VERIFYPEER , true);
+	} else {
```
Comment (alecpl) [inline plugins/password/drivers/httpapi.php:71]: We prefer `else` in a new line.

File: plugins/password/drivers/httpapi.php [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+	    $query = http_build_query($vars);
+	    if (parse_url($url, PHP_URL_QUERY)) { // Does URL already include a query string?
+		$url .= '&' . $query;
+	    }
+	    else {
+		$url .= '?' . $query;
+	    }
+	} else {
+	    rcube::raise_error(array(
+				     'code' => 600,
```
Comment (alecpl) [inline plugins/password/drivers/httpapi.php:96]: Again, not our style. It should be +8 spaces from the `rcube::` above.

File: plugins/password/drivers/httpapi.php [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+				     'type' => 'php',
+				     'file' => __FILE__, 'line' => __LINE__,
+				     'message' => "Password plugin: Invalid httpapi method"
+				     ), true, false);
+
+	    return PASSWORD_CONNECT_ERROR;
+	}
+
+        curl_setopt($curl, CURLOPT_URL            , $url);
```
Comment (alecpl) [inline plugins/password/drivers/httpapi.php:105]: Spacing.

File: plugins/password/README [hunk 6]
```diff
@@ -49,6 +49,7 @@
  2.1.21. Kpasswd
  2.1.22. Modoboa
  2.1.23. LDAP - Password Modify Extended Operation (ldap_exop)
+ 2.1.24. HTTP-API
```
Comment (alecpl) [inline plugins/password/README:52]: The driver name is `httpapi` so add `(httpapi)` in the section title.

**Review by alecpl — COMMENTED** [review]
I think that usefulness of this is very limited because of the lack of authentication. Is there any API that works this way? Could you add that?

**Comment by davidc** [comment]
Hi, thanks for these tips. The coding style is not mine, I tried to follow the style of the plesk driver I used as a template, but now realise that file wasn't actually in your own code style. I will correct these.

The reason for writing this was to write password changes to a simple API on our backend, but I decided to do it in a generic manner so it might be useful to other service providers, rather than requiring many similar or custom drivers.

The (simple) API on our end just verifies the host is the webmail server, and as it already checks that the user's current password is correct, this was thought sufficient in terms of authentication.

I will add in the ability to set custom headers and custom GET/POST variables (which can the be used for further authentication/API keys) in the hope that it is more universally useful.

**Comment by dereks** [comment]
Thank you for submitting a Pull Request (PR) to the Roundcube GitHub project.

You are receiving this message because your PR has conflicts which need to be resolved. We are trying to catch up on our backlog of old PRs and get them merged in (where appropriate). Therefor, we request the following:

**Step 1.** Rebase from the latest Roundcube branch master into your PR.

```
git fetch upstream
git checkout your_feature_branch_name
git rebase upstream/master
git push -f origin your_feature_branch_name
```

**Step 2.** Re-test to make sure your new code still works as expected

**Step 3.** Comment back here once it has been tested and will merge cleanly.

Once this has been done we will treat it like a new Pull Request and consider it for acceptance.

Apologies for the inconvenience. Thank you for contributing to Roundcube!

**Comment by alecpl** [comment]
Any progress with this? Since Roundcube uses Guzzle HTTP Client it would be nice to use it here too.

**Pull request merged**